VoiceController.php

```diff
@@ -1,48 +1,41 @@
-public function voice(Request $request){
-    $request->validate([
-        'question_id'=>'required|int|exists:questions,id',
-        'value'=>'required|boolean',
-    ]);
-
-    $question=Question::find($request->post('question_id'));
-    if (!$question)
-        return response()->json([
-            'status'=>404,
-            'message'=>'not found question ..'
-        ]);
-    if ($question->user_id==auth()->id())
-        return response()->json([
-            'status' => 500,
-            'message' => 'The user is not allowed to vote to your question'
-        ]);
+<?php
 
-    //check if user voted 
-    $voice=Voice::where([
-        ['user_id','=',auth()->id()],
-        ['question_id','=',$request->post('question_id')]
-    ])->first();
-    if (!is_null($voice)&&$voice->value===$request->post('value')) {
-        return response()->json([
-            'status' => 500,
-            'message' => 'The user is not allowed to vote more than once'
-        ]);
-    }else if (!is_null($voice)&&$voice->value!==$request->post('value')){
-        $voice->update([
-            'value'=>$request->post('value')
+// Controller
+class VoiceController
+{
+    public function voice(VoiceRequest $request)
+    {
+        $question = Question::findOrFail($request->question_id);
+
+        $question->voice()->updateOrCreate([
+            'user_id' => auth()->id(),
+            'value' => $request->value,
         ]);
+
         return response()->json([
-            'status'=>201,
-            'message'=>'update your voice'
+            'status' => 200,
+            'message' => 'Voting completed successfully',
         ]);
     }
 
-    $question->voice()->create([
-        'user_id'=>auth()->id(),
-        'value'=>$request->post('value')
-    ]);
+}
+
+// Form Request
+use Illuminate\Foundation\Http\FormRequest;
+
+class VoiceRequest extends FormRequest
+{
+    public function authorize()
+    {
+        return $this->user_id != auth()->id();
```

Comment:
@renwarkurd great move to use Form Request for this.

Reply:
There is no evidence that the request contains the user_id, and if it did, it could be forged.